### PR TITLE
Smooth Pistons

### PR DESCRIPTION
**Pull request by davchoo**
Fixes the flickering with pistons moving blocks. Needs GeyserMC/mappings#41
Fixes https://github.com/GeyserMC/Geyser/issues/2241

### CODE REVIEW

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:73]: There's shortcuts to these in the form of `Vector3i.UP`, `Vector3i.DOWN`... perhaps those would be cleaner here?

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:301]: Same deal as with the other comment.

**Review by DoctorMacc — COMMENTED** [review]
There's a lot to go through here - I'll try to run it later and get an IDE analysis of it.

**Comment by DoctorMacc** [comment]
https://cdn.discordapp.com/attachments/613194762249437245/777637784063180840/pistons-2020-11-15_15.53.02.mp4 there appears to be some buggy behavior.

**Comment by davchoo** [comment]
Ah, totally forgot about normal pistons.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (davchoo) [reply connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:73]: There's only shortcuts for `UP`, `RIGHT`, and `FORWARD`. I personally think mixing them with `Vector3i.from` may make it messier.

**Comment by DoctorMacc** [comment]
Solved on my end.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:415]: Do you ever update the player's motion to the server? Geyser doesn't do this automatically, you need to send a SetMotionPacket.

**Comment by DoctorMacc** [comment]
Looks good so far! I love that the slime block works correctly. A couple of critiques:

- I tested a [sticky] piston with no block attached and the current master behavior still applies. 
- Have you tried applying motion to the player in all cases instead of teleporting them? It might seem a little cleaner, and as of 1.16.20 I believe that is BDS behavior.

**Comment by zimzaza4** [comment]
It's great. It solved #1437

**Comment by davchoo** [comment]
I've fixed movement with pistons without any blocks attached. 

I didn't see any SetEntityMotion packets while using ProxyPass and BDS, but I tried it anyway. It kinda works but tends to push players sideways. You can see the effect when being launched by a slime block. 
I'm having trouble with the Java server teleporting players back to their original positions, which I believe causes honey and slime blocks to work intermittently.

**Comment by Hellohi3654** [comment]
after merging this PR with my fork I've been getting this 
image

**Comment by rtm516** [comment]
> after merging this PR with my fork I've been getting this

Can you try the build straight from this PR, as its likely its something else in your fork or an issue caused by a bad merge.

**Comment by Hellohi3654** [comment]
possibly doubt it though
I'll try

**Comment by Hellohi3654** [comment]
not the same error but I got a similar error while using the build straight from this PR
image
https://dump.geysermc.org/pHQveNjAFoioZxrVEo57lQ71il3AgCIn

**Comment by Hellohi3654** [comment]
the error is spamming

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (Heath123) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:386]: Shouldn't you set this back after?

**Review by Heath123 — CHANGES_REQUESTED** [review]

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (davchoo) [reply connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:386]: `playerCollision` is a separate object from the bounding box in `CollisionManager`.

**Comment by Striker872** [comment]
It doesn't work with chunk caching turned off.

**Comment by DoctorMacc** [comment]
@Striker872 correct; this PR relies on it being enabled, as do move fixes (which is partially used here to fix piston player movement).

File: connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaBlockValueTranslator.java
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaBlockValueTranslator.java:121]: If this something that needs to be addressed upstream?

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/BlockTranslator.java
Comment (DoctorMacc) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/BlockTranslator.java:271]: Can you move this code to BlockStateValues, along with `PISTON_HEADS`?

**Comment by Camotoy** [comment]
Is this ready for review?

**Comment by davchoo** [comment]
I don't think this PR is ready for review as there are still a few issues with this PR. Though I'm not sure how to fix them. 

- When players get launched sideways by a slime block, they sometimes get teleported back.
- When players get launched upwards, they get pushed horizontally off.
- When players are pushed sideways with fences/walls, it's sometimes treated as a full block.

I'm not sure of the cause of 1. I think 2 and 3 are caused by `movingBlock` having a full block collision.

**Comment by Heath123** [comment]
> `movingBlock` having a full block collision.
You can probablt chnage the regex in [EmptyCollsiion.java](https://github.com/GeyserMC/Geyser/blob/master/connector/src/main/java/org/geysermc/connector/network/translators/collision/translators/EmptyCollision.java) like with the shulker box in [SolidCollision.java](https://github.com/GeyserMC/Geyser/blob/c92150013f01a99ff36470eb67a7aa86d0ddb243/connector/src/main/java/org/geysermc/connector/network/translators/collision/translators/SolidCollision.java#L31)

**Comment by davchoo** [comment]
I assume that works based on the Java Identifier? I don't think `movingBlock` has a Java equivalent.

**Comment by zimzaza4** [comment]
It seems that some square l blocks will be dragged away, but it is not dragged away.
Screenshot_20201219_112134_com mojang minecraftpe

**Comment by zimzaza4** [comment]
It will move the block visually.

**Comment by zimzaza4** [comment]
Screenshot_20201219_113206_com mojang minecraftpe
Screenshot_20201219_113210_com mojang minecraftpe

In Java Edition, this block is not really pushed away

**Comment by davchoo** [comment]
@zimzaza4 The bug should be fixed in the latest commit.

**Comment by Heath123** [comment]
> I assume that works based on the Java Identifier? I don't think `movingBlock` has a Java equivalent.

`moving_piston` maybe?
Screenshot_20201219-203725_Chrome
https://minecraft.gamepedia.com/Piston

**Comment by Camotoy** [comment]
Don't forget to update copyrights to 2021!

**Comment by Hellohi3654** [comment]
What's the status on this PR?

**Comment by YHDiamond** [comment]
Usually the devs like people to fix merge conflicts before merging anything, I knew this from experience. I'd recommend davchoo does that but I don't get to decide anything.

**Comment by Camotoy** [comment]
Merge conflicts need to be fixed or we literally cannot merge it. That's why it's called a merge conflict.

As this PR is rather substantial, we just need to dedicate some time to properly reviewing it.

**Comment by Hellohi3654** [comment]
Yeah ok

**Comment by davchoo** [comment]
I'm also still working on fixing the teleports and glitchiness with slime blocks, which I think can be fixed by correcting the player's position before sending it to the Java server.

**Comment by YHDiamond** [comment]
build failed :(

**Comment by Heath123** [comment]
Wow, this is so much harder to add than I would have thought...

**Comment by Hellohi3654** [comment]
idk if you know this but
blocks go white when getting moved

**Comment by davchoo** [comment]
I'm aware. According to https://bugs.mojang.com/browse/MCPE-66250, it's a bug in Bedrock.

**Comment by Hellohi3654** [comment]
Oh ok it's just bugrock

**Comment by Camotoy** [comment]
What's the status of this PR? Is it ready for review?

**Comment by davchoo** [comment]
As of right now it mostly works. Slime blocks work pretty well. There are some issues with honey blocks. There are also still some glitches when pistons push players since I don't account for the direction of player movement and probably other issues in my code. I've pushed my local changes to smoothpistons2 that will probably resolve the last issue but currently breaks everything else.

**Comment by Hellohi3654** [comment]
Just to point out
When I merged this PR into my fork it caused movement issues
I'm not 100% sure if it was me when merging this PR but I did try merging it multiple times and every time it seemed to cause more lag backs

**Comment by Camotoy** [comment]
Status of this PR?

**Comment by Hellohi3654** [comment]
> Status of this PR?

Last time I merged this branch into my fork I had a few movement issues don't know if they have been fixed because I haven't remerged in a while

**Comment by davchoo** [comment]
I did some big changes to the movement in `fcbf537`, but I haven't experienced any movement issues outside of being pushed with a piston.

Here are some details about the current state of the PR.

Players still get teleported sometimes when being pushed with a piston. It causes slime blocks to sometimes not work and some movement glitches. I don't think I can improve on this any further.

Since #1814 is merged, I need to look into why sneaking & crawling won't activate when being crushed by a piston. 

I've disabled placing `movingBlock` block entities since they flash white due to https://bugs.mojang.com/browse/MCPE-66250. 

I don't think I need to do any major changes to the code, so you're welcome to review it if you want.

**Comment by Camotoy** [comment]
> Since #1814 is merged, I need to look into why sneaking & crawling won't activate when being crushed by a piston.

Perhaps that isn't defined by the server when done by a piston? We don't actively test for if the client should be crawling and sneaking, and I believe the Java Edition client does.

**Comment by Tydium** [comment]
Seems good to me 👍

File: connector/src/main/java/org/geysermc/connector/utils/BlockUtils.java
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/utils/BlockUtils.java:1]: This large diff is probably because I changed the line encoding from CRLF to LF, which is what I want to try to keep the line encoding as.

File: connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java
Comment (davchoo) [reply connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java:91]: Resetting needs to be done at the start because the flags in `PistonCache` is used between ticks in `CollisionManager::adjustBedrockPosition`. I've added an isEmpty check.

File: connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java
Comment (davchoo) [reply connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java]: It should only trigger if I break createMovingBlocks/removeMovingBlocks and leave PistonBlockEntity references in movingBlockMap.

File: connector/src/main/java/org/geysermc/connector/utils/BlockUtils.java
Comment (davchoo) [reply connector/src/main/java/org/geysermc/connector/utils/BlockUtils.java]: Strange, I thought git's `core.autocrlf` setting should prevent this from happening. It should be fixed now.

**Comment by Camotoy** [comment]
Climbing up stairs on vanilla 1.17.1 has definitely regressed with this PR - it now lags back.

**Comment by davchoo** [comment]
> At least in my dev environment, climbing up stairs on vanilla 1.17.1 has definitely regressed with this PR - it now lags back.

Should be fixed in the latest commit.

> If being launched by a slime block (or on top of any block), it appears that the piston does not show the slime block while in motion.

This is on purpose since the moving block causes players to be pushed to the side. See https://bugs.mojang.com/browse/MCPE-96035.

**Comment by Camotoy** [comment]
Definitely better, but it still can lag me back.

File: connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaBlockValueTranslator.java
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/network/translators/java/world/JavaBlockValueTranslator.java:74]: Set this to a static commit instead of just master; patch numbers tend to shift around quite a bit.

File: bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserPistonEvents.java
Comment (RednedEpic) [inline bootstrap/spigot/src/main/java/org/geysermc/platform/spigot/world/GeyserPistonEvents.java:53]: Should be renamed to be consistent with other listener names.
```suggestion
public class GeyserPistonListener implements Listener {
```

File: connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java:42]: Getter should be added to the top of this class since every variable (aside from GeyserSession) uses it. A getter with access level of none can be added to GeyserSession to prevent it from being accessible.
```suggestion
@Getter
```

File: connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/session/cache/PistonCache.java:107]: The `position` and `pistonBlockEntity` values can be inlined as they are only referenced once inside the log method.

File: connector/src/main/java/org/geysermc/connector/network/translators/collision/CollisionManager.java
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/translators/collision/CollisionManager.java:164]: The `0.08` value should be a constant.

File: connector/src/main/java/org/geysermc/connector/network/translators/collision/CollisionManager.java
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/translators/collision/CollisionManager.java:275]: The `0.6` value should be a constant.

**Review by RednedEpic — CHANGES_REQUESTED** [review]
Awesome work! Few nitpicks but generally most everything looks great.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:353]: If this method is only called for visuals, I would rather it just call the update block methods manually so we don't even touch the chunk cache.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:386]: Move this into the `if` statement below. Same with the next `for` loop.

File: connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/network/translators/world/block/entity/PistonBlockEntity.java:643]: Replace the Vector constructor with `Direction.DOWN.getUnitVector()`?

**Review by Camotoy — CHANGES_REQUESTED** [review]

**Review by Camotoy — APPROVED** [review]

**Pull request merged**
